lib/config.js

```diff
@@ -209,132 +90,275 @@ class Config {
          * If user declares "foo", convert to "foo:false".
          */
         this.globals = (options.globals || []).reduce((globals, def) => {
-            const parts = def.split(":");
+            const parts = def.split(SUBCONFIG_SEP);
 
             globals[parts[0]] = (parts.length > 1 && parts[1] === "true");
 
             return globals;
         }, {});
 
-        this.options = options;
-        this.loadConfigFile(options.configFile);
+        this.loadSpecificConfig(options.configFile);
+
+        // Empty values in configs don't merge properly
+        const cliConfigOptions = {
+            env: this.env,
+            rules: this.options.rules,
+            globals: this.globals,
+            parserOptions: this.parserOptions,
+            plugins: this.options.plugins
+        };
+
+        this.cliConfig = {};
+        Object.keys(cliConfigOptions).forEach(configKey => {
+            const value = cliConfigOptions[configKey];
+
+            if (value) {
+                this.cliConfig[configKey] = value;
+            }
+        });
     }
 
     /**
-     * Loads the config from the configuration file
-     * @param {string} configFile - patch to the config file
-     * @returns {undefined}
-     */
-    loadConfigFile(configFile) {
-        if (configFile) {
-            debug(`Using command line config ${configFile}`);
-            if (isResolvable(configFile) || isResolvable(`eslint-config-${configFile}`) || configFile.charAt(0) === "@") {
-                this.useSpecificConfig = loadConfig(configFile, this);
-            } else {
-                this.useSpecificConfig = loadConfig(path.resolve(this.options.cwd, configFile), this);
+    * Loads the config options from a config specified on the command line.
+    * @param {string} [config] A shareable named config or path to a config file.
+    * @returns {void}
+    */
+    loadSpecificConfig(config) {
+        if (config) {
+            debug(`Using command line config ${config}`);
+            const isNamedConfig =
+                isResolvable(config) ||
+                isResolvable(`eslint-config-${config}`) ||
+                config.charAt(0) === "@";
+
+            if (!isNamedConfig) {
+                config = path.resolve(this.options.cwd, config);
             }
+
+            this.useSpecificConfig = ConfigFile.load(config, this);
```

Comment:
Nitpick: Can you rename this property to something like `specificConfig` instead to clarify that it's a config object and not a boolean?